src/box/box.cc

```diff
@@ -6165,6 +6165,14 @@ box_shutdown(void)
 	 * for async watcher execution.
 	 */
 	box_watcher_shutdown();
+	/*
+	 * Iproto connections should be dropped before client fibers because
+	 * they can produce new ones in `tx_fiber_pool`.
+	 */
+	if (iproto_drop_connections(box_shutdown_timeout) != 0) {
+		diag_log();
+		panic("cannot gracefully shutdown iproto requests");
+	}
```

Comment:
> Currently we have such order of shutting down: client fibers, then iproto subsystem and then fiber pool (details can be found in https://github.com/tarantool/tarantool/commit/e513d4ce51d25ae56ca356183e6999ad6c298867). The problem is iproto produces new fibers in `tx_fiber_pool` when processing new requests so when Tarantool is shutting down client fiber, a new one can appear and it won't be handled. Moreover, there is no point in handling such fibers because one can appear even after all client fibers are finished.

A proper IPROTO connection should stop sending any requests and exit as soon as it returns a shutdown notification so no new fibers should appear from IPROTO when we proceed to cancelling client fibers. Can you describe in more details what happened in the failing tests?

Reply:
It seems that replication connection is not proper then - replica can send `IPROTO_SUBSCRIBE` even when master is shutting down, that's why it produces the new client fiber after all such fibers were cancelled. We could try to make it "proper", but it seems to be much more complicated than simply drop all iproto connections before finishing client fibers - then all connections will be handled correctly.